prompt2model/param_selector/search_with_optuna.py

```diff
@@ -0,0 +1,245 @@
+        """
+        self.generation_model_trainer = trainer
+        self.n_trials = n_trials
+
+    def optimize_hyperparameters(
+        self,
+        training_datasets: list[Dataset],
+        validation: Dataset,
+        hyperparameters: Optional[dict[str, Any]] = None,
+    ) -> dict[str, Any]:
+        """Select a model among a set of hyperparameters (given or inferred).
+
+        Args:
+            training_datasets (list[Dataset]): One or more training datasets or
+                subset of the actual dataset for running the trainer.
+            validation_sets (Dataset): A dataset for computing validation metrics.
+            hyperparameter_space (Optional[dict[str, Any]], optional): The set
+                of possible values of hyperparaneters values required for doing
+                optimal hyperparameter search. Defaults to None.
+
+        Returns:
+            Returns a dict which contains the best hyperparameters
+        """
+        supported_hp_space_keys = set(DEFAULT_HYPERPARAMETERS_SPACE.keys())
+        if hyperparameters is not None:
+            assert set(hyperparameters.keys()).issubset(
+                supported_hp_space_keys
+            ), f"Only support {supported_hp_space_keys} as training parameters."
+        hyperparameter_space = self._build_hp_space(hyperparameters)
+
+        concatenated_training_dataset = concatenate_datasets(training_datasets)
+        train_dataset = self.generation_model_trainer.tokenize_dataset(
+            concatenated_training_dataset
+        )
+
+        if isinstance(validation, list):
+            validation = concatenate_datasets(validation)
+            validation = self.generation_model_trainer.tokenize_dataset(validation)
+
+        def objective(trial: Trial):
+            model = self.generation_model_trainer.model
+            training_args = transformers.TrainingArguments(
+                output_dir="./checkpoint",
+                learning_rate=trial.suggest_loguniform(
+                    "learning_rate",
+                    low=hyperparameter_space["min_learning_rate"],
+                    high=hyperparameter_space["max_learning_rate"],
+                ),
+                weight_decay=trial.suggest_loguniform(
+                    "weight_decay",
+                    low=hyperparameter_space["min_weight_decay"],
+                    high=hyperparameter_space["max_weight_decay"],
+                ),
+                num_train_epochs=trial.suggest_int(
+                    "num_train_epochs",
+                    low=hyperparameter_space["min_num_train_epochs"],
+                    high=hyperparameter_space["max_num_train_epochs"],
+                ),
+                save_strategy=trial.suggest_categorical(
+                    "save_strategy", hyperparameter_space["save_strategy"]
+                ),
+                per_device_train_batch_size=trial.suggest_categorical(
+                    "per_device_train_batch_size",
```

Comment:
I feel like `per_device_train_batch_size` is not a good hyperparameter to tune, because you should generally just use the maximum batch size that fits in your machine's memory. I suggest removing this (and changing the default in `prompt2model/utils/config.py` to a constant)

Reply:
Yes, we can remove this. However, Suppose I have a batch size of 1, 2, 4. Where 4 is the max batch size. Now qualitative-wise (i.e. in terms of optimization), does greater batch size always provide better optimization? 

To explain with this example:

L1: aggregated validation loss for Batch 1
L2: aggregated validation loss for Batch 2
L3: aggregated validation loss for Batch 4

So is this thing correct always: L1 < L2 < L3. 

Because, we obviously know this, that resource for a batch size of 1 will be lesser than 2 and so on... Since we are doing optimization, I was thinking on that basis. Let me know what are your thoughts here.

Reply:
> So is this thing correct always: L1 < L2 < L3.

I don't think this is correct. Rather, in most cases, L1, L2, and L3 will be approximately identical.

The details are a little technical, but increasing the batch size from 1 to B reduces the variance of the stochastic gradient by 1/B (therefore reducing the number of optimization steps necessary by  a factor of 1/B). However, by increasing the batch size, we're now going to take |D| / B optimization steps rather than |D| steps (where D is the size of the dataset).

In theory, these will cancel out, and increasing the batch size on a fixed dataset will lead to a very similar trained model. In practice, this is desirable, because the time needed to compute a gradient on a batch (of size B) scales sublinearly with the batch size B, until you hit the memory limit of your machine. In practice, they also will likely yield similar performance.

Reply:
That's an awesome explanation. Thanks for the clarification.